dbms/src/Common/RWLock.cpp

```diff
@@ -53,6 +56,44 @@ class RWLockImpl::LockHolderImpl
 };
 
 
+namespace
+{
+    /// Global information about all read locks that query has. It is needed to avoid some type of deadlocks.
+
+    class QueryLockInfo
+    {
+    private:
+        std::mutex mutex;
+        std::map<std::string, size_t> queries;
+
+    public:
+        void add(const String & query_id)
+        {
+            std::lock_guard lock(mutex);
+            ++queries[query_id];
+        }
+
+        void remove(const String & query_id)
+        {
+            std::lock_guard lock(mutex);
+            auto it = queries.find(query_id);
+            assert(it != queries.end());
+            if (--it->second == 0)
+                queries.erase(it);
+        }
+
+        void check(const String & query_id)
+        {
+            std::lock_guard lock(mutex);
```

Comment:
`std::shared_lock()` would better fit in here as the method is **const** in its essence

Reply:
It looks more heavy for me.